### PR TITLE
CUDA: refactor topk-moe to enable more models (GLM 4.7, Nemotron etc.)

### PR DESCRIPTION
**Pull request by am17an**
Refactor the topk-moe to enabling various combination of topk-moe. Hopefully this will cover most models. I removed some templates from the code and only kept the bias because it has a extra warp shuffle, the rest of the template code does not provide any significant speedup.

3090

| Model                 | Test   |   t/s master |   t/s topk-cuda-refactor |   Speedup |
|:----------------------|:-------|-------------:|-------------------------:|----------:|
| deepseek2 ?B Q4_K_M   | pp512  |      3096.25 |                  3107.03 |      1.00 |
| deepseek2 ?B Q4_K_M   | tg128  |       124.49 |                   132.89 |      1.07 |
| gpt-oss 20B MXFP4 MoE | pp512  |      4773.87 |                  4751.48 |      1.00 |
| gpt-oss 20B MXFP4 MoE | tg128  |       211.17 |                   210.60 |      1.00 |
| qwen3moe 30B.A3B Q4_0 | pp512  |      3707.88 |                  3682.90 |      0.99 |
| qwen3moe 30B.A3B Q4_0 | tg128  |       190.88 |                   190.27 |      1.00 |

4090
| Model                 | Test   |   t/s master |   t/s topk-cuda-refactor |   Speedup |
|:----------------------|:-------|-------------:|-------------------------:|----------:|
| deepseek2 ?B Q4_K_M   | pp512  |      6447.25 |                  6503.07 |      1.01 |
| deepseek2 ?B Q4_K_M   | tg128  |       162.90 |                   175.63 |      1.08 |
| gpt-oss 20B MXFP4 MoE | pp512  |      9820.61 |                  9815.71 |      1.00 |
| gpt-oss 20B MXFP4 MoE | tg128  |       264.89 |                   264.37 |      1.00 |
| qwen3moe 30B.A3B Q4_0 | pp512  |      7827.77 |                  7821.77 |      1.00 |
| qwen3moe 30B.A3B Q4_0 | tg128  |       253.11 |                   252.62 |      1.00 |

5090
| Model                 | Test   |   t/s master |   t/s topk-cuda-refactor |   Speedup |
|:----------------------|:-------|-------------:|-------------------------:|----------:|
| deepseek2 ?B Q4_K_M   | pp512  |      6512.76 |                  6582.78 |      1.01 |
| deepseek2 ?B Q4_K_M   | tg128  |       170.66 |                   187.88 |      1.10 |
| gpt-oss 20B MXFP4 MoE | pp512  |     13815.72 |                 13798.87 |      1.00 |
| gpt-oss 20B MXFP4 MoE | tg128  |       384.08 |                   383.34 |      1.00 |
| qwen3moe 30B.A3B Q4_0 | pp512  |      7990.44 |                  8005.32 |      1.00 |
| qwen3moe 30B.A3B Q4_0 | tg128  |       285.53 |                   284.53 |      1.00 |

### CODE REVIEW

File: ggml/src/ggml-cuda/topk-moe.cu [hunk 1]
```diff
@@ -245,107 +329,76 @@ void ggml_cuda_op_topk_moe(ggml_backend_cuda_context & ctx,
...
+
     float clamp_val = -INFINITY;
-    if (with_norm) {
-        if (clamp) {
-            clamp_val = ggml_get_op_params_f32(clamp, 0);
-        }
-        launch_topk_moe_cuda<true>(ctx, logits_d, weights_d, ids_d, n_rows, n_experts, n_expert_used, clamp_val);
+    if (clamp) {
+        clamp_val = ggml_get_op_params_f32(clamp, 0);
+    }
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/topk-moe.cu:345]: Isn't the information stored in `with_norm` kind of redundant here? Is there a reason why the code couldn't or shouldn't check against `clamp_val == -INFINITY`?

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 2]
```diff
@@ -3080,63 +3080,163 @@ static bool ggml_cuda_should_fuse_rope_set_rows(const ggml_tensor * rope,
...
 
-    if (is_equal(topk_moe_ops_with_norm, ops) &&
-        ggml_can_fuse_subgraph(cgraph, node_idx, ops, { node_idx + 3, node_idx + 9 })) {
-        ggml_tensor * softmax = cgraph->nodes[node_idx];
-        ggml_tensor * weights = cgraph->nodes[node_idx + 9];
-        ggml_tensor * get_rows = cgraph->nodes[node_idx + 4];
-        ggml_tensor * argsort = cgraph->nodes[node_idx + 2];
-        int n_expert = cgraph->nodes[node_idx]->src[0]->ne[0];
+    if (nodes[node_idx]->op == GGML_OP_UNARY) {
+        args.sigmoid = true;
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/ggml-cuda.cu:3101]:
```suggestion
        GGML_ASSERT(ggml_get_unary_op(nodes[node_idx]) == GGML_UNARY_OP_SIGMOID);
        args.sigmoid = true;
```

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 3]
```diff
@@ -3080,63 +3080,163 @@ static bool ggml_cuda_should_fuse_rope_set_rows(const ggml_tensor * rope,
...
+        args.delayed_softmax = true;
     }
 
-    if (is_equal(topk_moe_ops, ops) && ggml_can_fuse_subgraph(cgraph, node_idx, ops, { node_idx + 3, node_idx + 4 })) {
-        ggml_tensor * softmax = cgraph->nodes[node_idx];
-        ggml_tensor * weights = cgraph->nodes[node_idx + 4];
-        ggml_tensor * get_rows = cgraph->nodes[node_idx + 4];
-        ggml_tensor * argsort = cgraph->nodes[node_idx + 2];
-        int n_expert = cgraph->nodes[node_idx]->src[0]->ne[0];
+    node_idx++;
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/ggml-cuda.cu:3108]: Generally speaking I'm not a fan of changing function arguments in the body (though in this case I think it's still fine).

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 4]
```diff
@@ -3398,35 +3498,75 @@ static void ggml_cuda_graph_evaluate_and_capture(ggml_backend_cuda_context * cud
                 // start of fusion operations
                 static bool disable_fusion = (getenv("GGML_CUDA_DISABLE_FUSION") != nullptr);
                 if (!disable_fusion) {
+                    ggml_cuda_topk_moe_args args;
+
+                    if (cgraph->nodes[i]->op == GGML_OP_UNARY || cgraph->nodes[i]->op == GGML_OP_SOFT_MAX ||
+                        cgraph->nodes[i]->op == GGML_OP_ARGSORT) {
+                        bool can_fuse = ggml_cuda_topk_moe_fusion(cgraph, i, args);
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/ggml-cuda.cu]:
```suggestion
                        const bool can_fuse = ggml_cuda_topk_moe_fusion(cgraph, i, args);
```

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 5]
```diff
@@ -3080,63 +3080,163 @@ static bool ggml_cuda_should_fuse_rope_set_rows(const ggml_tensor * rope,
...
-        ggml_tensor * argsort = cgraph->nodes[node_idx + 2];
-        int n_expert = cgraph->nodes[node_idx]->src[0]->ne[0];
+    node_idx++;
 
-        if (ggml_cuda_should_use_topk_moe(softmax, weights, get_rows, argsort, nullptr, n_expert)) {
-            return true;
+    if (args.sigmoid || args.softmax) {
+        // SOFTMAX -> RESHAPE
+        if (node_idx >= n_nodes || nodes[node_idx]->op != GGML_OP_RESHAPE ||
+            nodes[node_idx]->src[0] != nodes[node_idx - 1]) {
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/ggml-cuda.cu]:
```suggestion
                nodes[node_idx]->src[0] != nodes[node_idx - 1]) {
```

I think it makes sense to indent by 8 spaces rather than 4 to make the logic visually distinct from the code that would oftentimes follow on this line.

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 6]
```diff
@@ -3398,35 +3498,75 @@ static void ggml_cuda_graph_evaluate_and_capture(ggml_backend_cuda_context * cud
...
-                        ggml_tensor * ids     = cgraph->nodes[i + 1];
+                                weights      = cgraph->nodes[i + ops.size() - 1];
+                                out_nodes[1] = i + ops.size() - 1;
 
-                        ggml_cuda_op_topk_moe(*cuda_ctx, node->src[0], weights, ids, /*with norm*/ false,
-                                              /*delayed_softmax*/ true);
-                        i += 5;
-                        continue;
+                                if (ggml_can_fuse_subgraph(cgraph, i, ops.size(), ops.data(), out_nodes, 2) &&
+                                    ggml_cuda_should_use_topk_moe(node, logits, weights, ids)) {
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/ggml-cuda.cu]:
```suggestion
                                        ggml_cuda_should_use_topk_moe(node, logits, weights, ids)) {
```

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 7]
```diff
@@ -3398,35 +3498,75 @@ static void ggml_cuda_graph_evaluate_and_capture(ggml_backend_cuda_context * cud
...
+                                //special case gpt-oss, no norm, no bias.
+                                ops.insert(ops.end(), { GGML_OP_ARGSORT, GGML_OP_VIEW, GGML_OP_GET_ROWS,
+                                                        GGML_OP_RESHAPE, GGML_OP_SOFT_MAX, GGML_OP_RESHAPE });
+                                weights                     = cgraph->nodes[i + 5];
+                                ids                         = cgraph->nodes[i + 1];
+                                const ggml_tensor * softmax = cgraph->nodes[i + 4];
+
+                                int out_nodes[2] = { i + 1, i + 5 };
+                                if (ggml_can_fuse_subgraph(cgraph, i, ops.size(), ops.data(), out_nodes, 2) &&
+                                    ggml_cuda_should_use_topk_moe(softmax, logits, weights, ids)) {
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/ggml-cuda.cu]:
```suggestion
                                        ggml_cuda_should_use_topk_moe(softmax, logits, weights, ids)) {
```

**Review by JohannesGaessler — APPROVED** [review]

File: ggml/src/ggml-cuda/ggml-cuda.cu (hunk 2 above)
Comment (am17an) [reply ggml/src/ggml-cuda/ggml-cuda.cu:3101]: actually better would be to return rather than assert

File: ggml/src/ggml-cuda/ggml-cuda.cu (hunk 5 above)
Comment (am17an) [reply ggml/src/ggml-cuda/ggml-cuda.cu]: I ran this though clang-format and it does this, though I agree with you.

File: ggml/src/ggml-cuda/topk-moe.cu (hunk 1 above)
Comment (am17an) [reply ggml/src/ggml-cuda/topk-moe.cu:345]: you're right. This can be simplified
